core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -209,14 +209,20 @@ public void onFailure(Exception replicaException) {
                     shardReplicaFailures.add(new ReplicationResponse.ShardInfo.Failure(
                         shard.shardId(), shard.currentNodeId(), replicaException, restStatus, false));
                     String message = String.format(Locale.ROOT, "failed to perform %s on replica %s", opType, shard);
+                    // TODO: Boaz wanted to get rid of this warning, should we?
                     logger.warn(
                         (org.apache.logging.log4j.util.Supplier<?>)
                             () -> new ParameterizedMessage("[{}] {}", shard.shardId(), message), replicaException);
-                    replicasProxy.failShard(shard, replicaRequest.primaryTerm(), message, replicaException,
-                        ReplicationOperation.this::decPendingAndFinishIfNeeded,
-                        ReplicationOperation.this::onPrimaryDemoted,
-                        throwable -> decPendingAndFinishIfNeeded()
-                    );
+                    final boolean shardWasFailed = replicasProxy.failShardIfNeeded(shard, replicaRequest.primaryTerm(), message,
```

Comment:
I'd rather not introduce more code paths. I would vote to always have the callback pattern. If people have nothing to do they can call onSuccess on the same thread.